srml/staking/src/phragmen.rs

```diff
@@ -207,162 +212,180 @@ pub fn elect<T: Trait + 'static, FV, FN, FS>(
-	_tolerance: BalanceOf<T>
-) -> BalanceOf<T> {
-	let into_currency = |b: BalanceOf<T>|    <T::CurrencyToVote as Convert<BalanceOf<T>, u64>>::convert(b) as ExtendedBalance;
-	let into_votes = |b: ExtendedBalance| <T::CurrencyToVote as Convert<ExtendedBalance, BalanceOf<T>>>::convert(b);
-	let tolerance = into_currency(_tolerance);
-
-	let mut elected_edges = nominator.edges
-		.iter_mut()
-		.filter(|e| e.elected)
-		.collect::<Vec<&mut Edge<T::AccountId>>>();
-
-	if elected_edges.len() == 0 {
-		return <BalanceOf<T>>::zero();
+	assignments: &mut Vec<(T::AccountId, BalanceOf<T>, Vec<Assignment<T>>)>,
+	expo_map: &mut ExpoMap<T>,
+	tolerance: ExtendedBalance,
+	iterations: usize,
+) {
+	for _i in 0..iterations {
+		let mut max_diff = 0;
+		assignments.iter_mut().for_each(|(n, budget, assignment)| {
+			let diff = do_equalize::<T>(&n, *budget, assignment, expo_map, tolerance);
+			if diff > max_diff {
+				max_diff = diff;
+			}
+		});
+		if max_diff < tolerance {
+			break;
+		}
 	}
+}
+
+fn do_equalize<T: Trait + 'static>(
+	nominator: &T::AccountId,
+	budget_balance: BalanceOf<T>,
+	elected_edges_balance: &mut Vec<Assignment<T>>,
+	expo_map: &mut ExpoMap<T>,
+	tolerance: ExtendedBalance
+) -> ExtendedBalance {
+	let to_votes = |b: BalanceOf<T>| <T::CurrencyToVote as Convert<BalanceOf<T>, u64>>::convert(b) as ExtendedBalance;
+	let to_balance = |v: ExtendedBalance| <T::CurrencyToVote as Convert<ExtendedBalance, BalanceOf<T>>>::convert(v);
+	let budget = to_votes(budget_balance);
+
+	// Convert all stakes to extended. Result is Vec<(Acc, Ratio, Balance)>
+	let mut elected_edges = elected_edges_balance
+		.into_iter()
+		.map(|e| (e.0.clone(), e.1, to_votes(e.2)))
+		.collect::<Vec<(T::AccountId, ExtendedBalance, ExtendedBalance)>>();
 
 	let stake_used = elected_edges
 		.iter()
-		.fold(0, |s, e| s.saturating_add(e.backing_stake));
-	let backed_stakes = elected_edges
+		.fold(0 as ExtendedBalance, |s, e| s.saturating_add(e.2));
+
+	let backed_stakes_iter = elected_edges
 		.iter()
-		.map(|e| elected_candidates[e.elected_idx].backing_stake)
-		.collect::<Vec<ExtendedBalance>>();
+		.filter_map(|e| expo_map.get(&e.0))
+		.map(|e| to_votes(e.total));
+
 	let backing_backed_stake = elected_edges
 		.iter()
-		.filter(|e| e.backing_stake > 0)
-		.map(|e| elected_candidates[e.elected_idx].backing_stake)
+		.filter(|e| e.2 > 0)
+		.filter_map(|e| expo_map.get(&e.0))
+		.map(|e| to_votes(e.total))
 		.collect::<Vec<ExtendedBalance>>();
 
 	let mut difference;
 	if backing_backed_stake.len() > 0 {
-		let max_stake = *backing_backed_stake
+		let max_stake = backing_backed_stake
 			.iter()
 			.max()
 			.expect("vector with positive length will have a max; qed");
-		let min_stake = *backed_stakes
-			.iter()
+		let min_stake = backed_stakes_iter
 			.min()
-			.expect("vector with positive length will have a min; qed");
+			.expect("iterator with positive length will have a min; qed");
```

Comment:
Yes, but you did not check that `backed_stakes_iter.len() > 0` before entering this block. Seems that to be certain you'd have to check that the input `elected_edges_balance` is longer than zero before passing it to this function, so probably best to put the check within the function, no?

Reply:
So `backed_stakes_iter` is a superset of `backing_backed_stake` (see the filter applied to both) which is checked to have a positive length. I believe this is safe enough to not add another condition.